File: js-lang/src/nodes/type.ts
import { Scope, Symbol } from "../common";
import { Type } from "../types/common";
import { numberType } from "../types";
import { stringType } from "../types";
import { unknownType } from "../types/unknown";
import { AnalyzeContext, Node } from "./base";
import { IdentifierNode } from "./identifier";
import { LiteralNode } from "./literal";

export class TypeNode extends Node {
  private _type: Type = unknownType;

  public get type() {
    return this._type;
  }

  private typeSymbol: Symbol | null = null;

  public setupScopes(scope: Scope, context: AnalyzeContext) {
    if (this.node instanceof IdentifierNode) {
      this.typeSymbol = scope.resolveLast(this.node.token.text);
      // TODO report if not found
    } else {
      this._type = {
        kind: "Literal",
        value: this.node.value,
        valueType:
          typeof this.node.value === "string" ? stringType : numberType,
      };
    }

    this.scope = scope;
  }

  public analyze(context: AnalyzeContext) {
    if (this.typeSymbol) {
      this.typeSymbol.analyze(context);
      this._type = this.typeSymbol.valueType;
    }
  }

  constructor(public readonly node: LiteralNode | IdentifierNode) {
    super([node]);
  }
}

File: dotnet/Tybscri/Nodes/Type.cs
using System.Linq.Expressions;

namespace Tybscri.Nodes;

public abstract class TypeNode : Node
{
    public abstract TybscriType Type { get; }
    
    public override Expression ToClrExpression(GenerateContext generateContext)
    {
        throw new NotSupportedException("A type node cannot be converted to an LINQ expression");
    }
}

public class LiteralTypeNode : TypeNode
{
    public LiteralTypeNode(TybscriType type)
    {
        Type = type;
    }

    public override TybscriType Type { get; }
}

public class IdentifierTypeNode : TypeNode
{
    private Symbol? _symbol;
    
    public IdentifierNode Identifier { get; }

    public IdentifierTypeNode(IdentifierNode identifier)
    {
        Identifier = identifier;
    }

    public override void ResolveTypes(CompileContext context, AnalyzeContext analyzeContext)
    {
        _symbol = Scope.ResolveLast(Identifier.Name);
        _symbol?.ResolveTypes(context);
    }

    public override TybscriType Type => _symbol?.ValueType ?? StandardTypes.Unknown;
}

//
//
// public class TypeNode : Node {
//     public Token? Literal { get; }
//     public IdentifierNode? Identifier { get; }
//
//     private Symbol? typeSymbol;
//
// public TypeNode(Token literal) : base()
// {
//     Literal = literal;
// }
//
// public TypeNode(IdentifierNode identifier) : base()
// {
//     Identifier = identifier;
// }
//
// public Scope SetupScopes(Scope scope) {
//     if (Identifier is not null) {
//         this.typeSymbol = scope.GetLast(Identifier.Name);
//         // TODO report if not found
//     } else {
//         switch (this.Literal) {
//             { Valu}
//         }
//         _type = new BooleanLiteralType() {
//             kind: "Literal",
//             value: this.node.value,
//             valueType:
//             typeof this.node.value === "string" ? stringType : numberType,
//         };
//     }
//
//     this.scope = scope;
// }
//
// public resolveTypes(context: CompileContext) {
//     if (this.typeSymbol) {
//         this.typeSymbol.resolveTypes(context);
//         this._type = this.typeSymbol.valueType;
//     }
// }
// }
